src/content/blog/val-vibes.mdx

```diff
@@ -0,0 +1,166 @@
+import Indexing3 from "./val-vibes/indexing3.png";
+import Search1 from "./val-vibes/search1.png";
+import Search2 from "./val-vibes/search2.png";
+import Blobindex from "./val-vibes/blobindex.png";
+import Blobsearch from "./val-vibes/blobsearch.png";
+import ManLicksDog from "./val-vibes/man-licks-dog.jpg";
+import Prod from "./val-vibes/prod.png";
+
+We've launched a new feature: semantic search! In this post we'll look at what it is, how we prototyped it, and what's next.
+
+### Similar vibes
+
+“Search” is finding documents based on a query, using some sort of sorting mechanism that ranks the most relevant documents. One such sorting mechanism is “semantic search”, which uses one of those crazy features to come out of the world of machine learning: embeddings. It works like this: you put in a string into an API, and you get an array of numbers (the “embedding”), with the property that things that have the same “vibe” will get similar numbers.
+
+For example, “animal that barks” should have similar vibes to “dog”. And “dog bites man” should have different vibes than “man bites dog”. And “discord bot” should have _very_ different vibes still.
+
+<figure style={{ maxWidth: "300px" }}>
+  <Image src={ManBitesDog} alt="Man bites dog" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+You generate embeddings for all your documents (ahead of time), as well as for the search query, and then you return documents sorted by how similar the embeddings are.
+
+That's the idea, though in practice embeddings can be quite unpredictable and surprising. Our examples above already break down… “animal that barks” apparently vibes more with “woof woof woof” than with “dog”. And “dog bites man” and “man bites dog” have even more similar vibes than that, presumably because there's lots of biting going on in both cases.
+
+For our purposes though, embedding similarity is high between “discord bot” and “github api”, or “chat server integration”. This helps in finding code that is not exactly the same, but still very relevant to your search query.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={CompareEmbeddings} alt="" />
+  <figcaption>
+    Woof woof woof.
+  </figcaption>
+</figure>
+
+At Val Town we had existing search for executable code snippets (“vals”), but it did only exact matches, often returning only a few results, or none at all. Surely semantic search would help return vals with similar vibes, despite getting confused by barking and biting.
+
+### Prototyping in Val Town
+
+I ended up making three different prototypes, using different storage engines: [Postgres](https://www.postgresql.org/) (using [Neon](https://neon.tech/)), [SQLite](https://sqlite.org/) (using [Turso](https://turso.tech/)), and in-memory (using [blob storage](https://docs.val.town/std/blob/)). Let's look at Postgres first, even though I built it last, since it's the simplest. At the end I'll also show how to do this in-memory.
```

Comment:
```suggestion
I ended up making three different prototypes, using different storage engines: [Postgres](https://www.postgresql.org/) (using [Neon](https://neon.tech/)), [SQLite](https://sqlite.org/) (using [Turso](https://turso.tech/)), and in-memory (using [blob storage](https://docs.val.town/std/blob/)). Let's look at Postgres first, and at the end I'll also show how to do this in-memory.
```

I don't think the ordering/difficulty is necessary here and makes the sentence a little confusing.